Add Mean and Median to std.algorithm

JackStouffer opened the pull request:
This covers [issue 14034](https://issues.dlang.org/show_bug.cgi?id=14034).

I couldn't find a way to make a median function for input ranges that didn't use the GC, so I just left it to random access ranges.

yebblies commented on std/algorithm/iteration.d:
```diff
@@ -35,6 +35,10 @@ $(T2 joiner,
         $(D map!"2 * a"([1, 2, 3])) lazily returns a range with the numbers
         $(D 2), $(D 4), $(D 6).)
+$(T2 mean,
+        Lazily finds the mean of a range)
```
This function does not find the mean lazily.

yebblies commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4284,77 @@ unittest
+auto mean(R)(R r) if (isInputRange!R && !isInfinite!R) {
+    import std.typecons : tuple;
+
+    immutable auto pair = reduce!((a, b) => tuple(a[0] + 1, a[1] + b))(tuple(0u, 0.0), r);
```
This is not going to work well for reals, longs, any ints where the sum overflows, etc
Should probably use `sum`.
Wastes a lot of time re-calculating the length if the range has it available.

yebblies commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4284,77 @@ unittest
+
+///
+@safe unittest {
+    assert(mean([1, 2, 3]) == 2);
```
Using `==` with floating point is not a great idea.

JackStouffer replied:
Do you mean there would be a problem with floating point imprecision? How else do you check for equality?

yebblies commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4284,77 @@ unittest
+        return r[0];
+    }
+
+    auto sorted = sort(r.save);
```
This does not mean the original data is unchanged.  You need to at least document that the data will be modified.

JackStouffer replied:
Would it be better to make a copy of the data and then sort it, or would it be better to just have the function mutate the range?

yebblies replied:
It might be better to make median only work on already sorted ranges, so the sort in user code is explicit.

yebblies commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4284,77 @@ unittest
+
+    if (sorted.length % 2 == 0)
+    {
+        return mean([sorted[($ / 2) - 1], sorted[$ / 2]]);
```
Better to just add and divide by two, or use `std.range.only`.

JackStouffer commented:
Fixed mentioned issues and rebased.

quickfur commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4284,111 @@ unittest
+Returns:
+    the mean of r
+*/
+auto mean(R)(R r) if (isInputRange!R && !isInfinite!R)
```
There's no check in the sig constraints for the range elements to be numeric types, yet the function freely applies `+` and `0.0` to the elements.

quickfur replied:
I.e., the sig constraints should at least check that range elements are convertible to floating point (which precision, though?), otherwise `+` and `0.0` will cause a compile error.

JackStouffer replied:
Would
```
isImplicitlyConvertible!(ElementType!R, float) || isImplicitlyConvertible!(ElementType!R, double)
```
work?

quickfur replied:
That's a lot of typing (and reading). What about just straight-up:
````
is(ElementType!R : double)
````
?

For simplicity, I'd just stick with `double` to minimize precision errors. But if you *really* wanted to, you could parametrize the floating type and use `isFloatingPoint!F` in the sig constraints to enforce it, then the user can specify the type to use for computing the average.

quickfur commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4284,111 @@ unittest
+    {
+        import std.typecons : tuple;
+
+        immutable auto pair = reduce!((a, b) => tuple(a[0] + 1, a[1] + b))(tuple(0u, 0.0), r);
```
Is `immutable` here really necessary?

Also, this line is a bit on the long side. The preferred line length limit (not a hard limit, though) is 80 columns.

e10s replied:
> `tuple(0u, 0.0)`

`0u` is not enough. How about `size_t(0)` or `cast(size_t)0`?

quickfur commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4284,111 @@ unittest
+        return (r[0] + r[1]) / 2;
+    }
+
+    auto sorted = assumeSorted(r);
```
Uhm, this is wrong. `assumeSorted` should be in *user code*, not here. This function should take `SortedRange` as parameter, so that the user has to either sort his range, or call `assumeSorted`, before this function will accept the range. Putting `assumeSorted` here completely defeats the purpose of `SortedRange`.

quickfur commented:
BTW, it is possible to implement `median` for (sorted) forward ranges without allocating. If `length` is defined, you just drop the first `r.length/2` elements of the range and compute the median.

If `length` is not defined, `save` the forward range at the beginning, then iterate over the original range and the `save`d range at half the speed, i.e., for every 2 elements popped from the original range, pop 1 element from the `save`d range. When the original range is exhausted, the `save`d range will be either on the median element, or on the element right before the median, and can therefore be computed easily.

For bidirectional ranges, you can just call `popFront` and `popBack` until the range is down to just 1 or 2 elements. (If you save the values of the last front and back elements, you can just pop both ends until empty then return their average.)

Note, though, that for anything other than a random access range with `length`, the complexity of `median` will be O(n).  This should be, of course, documented clearly. Also, in all cases `median` should take a `SortedRange`, so that the onus is on the caller to sort his data (or ensure it arrives in sorted order) before feeding it to `median`.

quickfur commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4284,111 @@ unittest
+}
+
+///
+@safe unittest {
```
Hate to be nitpicking on this again, but Phobos style guide requires opening braces to be on a new line.

quickfur commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4284,111 @@ unittest
+    assert(approxEqual(mean([1.5, 2.5, 12.5]), 5.5));
+}
+
+unittest {
```
ditto

quickfur commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4284,111 @@ unittest
+}
+
+///
+@safe unittest {
```
ditto

JackStouffer replied:
Sorry, keep forgetting. I'll need to download dfmt and setup a phobos style to make sure I don't do this again.

e10s commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4284,111 @@ unittest
+    static if (hasLength!R)
+    {
+        assert(r.length > 0);
+    }
```
Is `assert(!r.empty)` necessary and sufficient in all cases?

JackStouffer commented:
@quickfur Addressed all concerns and rebased.

JackStouffer commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4285,307 @@ unittest
+    //{
+    //    return r.popFront;
+    //}
+
```
As it states in the comment, I don't get this.

quickfur replied:
Because `popFront` does not return anything. Don't you mean `r.front` instead?

JackStouffer replied:
Yes, your right.

burner commented on std/algorithm/iteration.d:
```diff
@@ -62,6 +66,7 @@ module std.algorithm.iteration;
 // FIXME
 import std.functional; // : unaryFun, binaryFun;
+import std.range;  // : SortedRange;
 import std.range.primitives;
```
if you import all of std.range why do you still import std.range.primitives ?

only import what you really need.

burner commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4285,307 @@ unittest
+    import std.math : approxEqual;
+
+    assert(mean([1, 2, 3]) == 2);
+    assert(approxEqual(mean([1.5, 2.5, 12.5]), 5.5));
```
test with empty range (see previous comment)

JackStouffer commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4285,307 @@ unittest
+    is(ElementType!R : double))
+in
+{
+    assert(!r.empty);
```
But the mean of an empty range isn't zero, it's nothing. I feel that if an empty range is passed to this function, it's more likely than not that a mistake was made.

burner replied:
then throw an exception, asserts may be ignored by the compiler

JackStouffer replied:
But wouldn't that cause this function to not be marked as @nogc

burner replied:
yes, but better than be maybe incorrect

JackStouffer replied:
I was under the impression that all functions in std.algorithm needed to not allocate GC memory.

burner replied:
maybe, but an algorithm that is sometimes wrong depending on the compiler flags is worse.

maybe return double.nan for zero length ranges

JackStouffer replied:
Ok, that sounds good. I will still leave in the contract though to help people debug

DmitryOlshansky replied:
> Ok, that sounds good. I will still leave in the contract though to help people debug

Sensible thing to do. We generally assert on things like popFront of an empty range due to performance considerations.

JackStouffer replied:
Wouldn't `assert(r.popFront);` cause the first element of the range to be dropped before the function is executed?

DmitryOlshansky replied:
What I mean is :
`void popFront(){ assert(!empty); .. }`

So it's inside of popFront not on popFront sorry.

burner commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4285,307 @@ unittest
+    assert(r7.assumeSorted.median == 5);
+
+    DummyRange!(ReturnBy.Reference, Length.Yes, RangeType.Bidirectional) r8;
+    assert(r8.assumeSorted.median == 5);
```
this could be a bit more DRY

```d
foreach(RV; TypeTuple!(ReturnBy.Reference,ReturnBy.Value)) {
    foreach(L; TypeTuple!(Length.No,Length.Yes)) {
    ....
    }
}
```

burner commented:
as the median is only a fancy word for the 0.5 quantil why not add a generic quantil functions like:

```d
auto quantil(R)(R r, double q) { // q <= 0.0 && q <= 1.0
```
?

JackStouffer commented:
And then add a convenience function for the median? Yeah that sounds reasonable.

JackStouffer commented:
Changed code to match @burner's requests.

I'm not happy with the `cast`'s everywhere, but I don't think there is another way to do it.

quickfur commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4284,237 @@ unittest
+    }
+    else
+    {
+        return r[cast(ulong) j];
```
Are you sure you should be using casts here? What about using `std.math.round` or `std.math.trunc`?

JackStouffer replied:
Fixed

JackStouffer commented:
The submitted code is incorrect as `long` cannot be used as array indexes on 32 bit systems. Other than using casts, I see no way to round a `real` to an `int` because all of the rounding functions give `long`.

JackStouffer commented:
Updated code and auto tester is green.

burner commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4284,265 @@ unittest
+    }
+
+    real j;
+    real g;
```
real g = modf(x,f);

burner commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4284,265 @@ unittest
+    g = modf(x, j);
+    uint j_int = cast(uint) j;
+
+    if (g == 0)
```
comparing real for equality is bad. maybe:

cast(int)g == 0

burner replied:
the second version has the same issue

JackStouffer replied:
Fixed, but instead of using a cast there which marked the function as unsafe, I used `approxEqual`.

burner commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4284,263 @@ unittest
+    double x = r.length * percentile;
+    real j;
+    real g = modf(x, j);
+    uint j_int = cast(uint) j;
```
the lines between 4383 to 4397 are equal between this functions and the next. maybe you can merge this to make it more DRY

burner commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4284,263 @@ unittest
+    assert([2.9, 6.2, 6.5, 20.5].assumeSorted.median.approxEqual(6.35));
+    assert([2.9, 5.8, 6.2, 6.5, 20.5].assumeSorted.median.approxEqual(6.2));
+
+    assert([12, 5, 3].assumeSorted!"a > b".median == 5);
```
please test empty ranges as well

JackStouffer replied:
Is there a way to do something like this in D?

```
debug
{
    //only in debug
}
else
{
    //only in release
}
```
I ask because I need two seperate tests for the debug and release versions.

DmitryOlshansky replied:
> I ask because I need two seperate tests for the debug and release versions.

I'd suggest just adding a bunch of asserts in debug mode since nobody is expected to compile as: `dmd -debug -release abc.d` ;)

burner commented:
version(debug) {
} else {
}

but having test behavior change between debug and release IMO is wron

JackStouffer commented:
The test behavior needs to change for the empty range test as the contract will fail in debug mode but in release mode the function will return `double.nan`.

JackStouffer commented:
@burner your solution doesn't compile

```
std/algorithm/iteration.d(4358): Error: identifier or integer expected, not debug
```

DmitryOlshansky commented:
> debug mode but in release mode the function will return double.nan.

Just always return nan and in debug mode it will assert before that?

JackStouffer commented:
> Just always return nan and in debug mode it will assert before that?

Did you accidentally a word?

JackStouffer commented:
`quantile` should be pure, but is isn't because of its use of `modf`. I submitted a pull at https://github.com/D-Programming-Language/druntime/pull/1373 which would automatically mark `quantile` as pure.

JackStouffer commented:
Adressed all comments and rebased.

quickfur commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4284,238 @@ unittest
+Returns:
+    the quantile of the range given $(D percentile)
+*/
+double quantile(alias less = "a < b", R)(SortedRange!(R, less) r, double percentile)
```
Does this code actually work for any other ordering predicate than `a < b`??

quickfur replied:
Also, are you sure `percentile` is the right name for this parameter? Wikipedia says that `percentile` means the 100th quantile.

Also, what happens if the user passes in a value outside of the range 0.0 to 1.0? I.e., this requirement should be stated, and enforced in the in-contract.

JackStouffer replied:
Yes, see https://github.com/D-Programming-Language/phobos/pull/3592/files#diff-0d6463fc6f41c5fb774300832e3135f5R4499

quickfur replied:
Does `quantile!"a > b"(r, 0.25)` return the right answer?  What if I call `quantile!"a**2 < b**2"`?

JackStouffer replied:
I believe you are misinterpreting what wikipedia is saying. A percentile is not `quantile(1)` it's the range broken into 100 different pieces. e.g. saying the 99th percentile means that 99% of the numbers in the range are below that number.

quickfur replied:
Yes, so calling this parameter `percentile` is misleading, because if I pass in, say, `50`, it won't give me the median. Instead, I have to pass in `0.5`, which, according to the definition of percentile, should mean 0.5%, not 50%.

I concede that there doesn't seem to be a better name for this parameter, though. I would just call it something like `q` and then explain what it means in the docs.

quickfur commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4284,238 @@ unittest
+        {
+            foreach (index, item; r.enumerate)
+            {
+                if (index == j_int)
```
Why not just replace the loop with `r.drop(j_int)`?

quickfur commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4284,238 @@ unittest
+                {
+                    return (item + prev_item) / 2.0;
+                }
+            }
```
Probably can replace the loop with `r.drop(j_int - 1)`?

JackStouffer commented:
Adressed all concerns and rebased. Also changed ddoc to use proper big O notation.

quickfur commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4284,247 @@ unittest
+
+Params:
+    r = a forward range with length or a random access range
+    q = where to calculate the quantile
```
Should indicate that the value is between 0.0 and 1.0, and explain what that means (0.0 = at the lowest end, 1.0 = at the highest end, etc.).

quickfur commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4284,247 @@ unittest
+    real g = modf(x, j);
+    uint j_int = cast(uint) j;
+
+
```
Nitpick: extraneous blank line here.

quickfur commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4284,247 @@ unittest
+Returns:
+    the quantile of the range given $(D q)
+*/
+double quantile(alias less = "a < b", R)(SortedRange!(R, less) r, double q)
```
There is still the concern over what the `less` parameter means, in terms of the returned value. What is the result of `r.quantile!`a > b`(0.25)`, for example? What does it mean, numerically? What *should* it mean?

What I'm trying to get at, is, does it even make any sense for `less` to be anything other than `a < b`? If `less` is any other ordering than `a < b`, will this function actually compute the quantile, or will it compute something completely different? If it doesn't make sense for `less` to be anything other than `a < b`, why is it a parameter that can be specified by the user?

yebblies replied:
It should probably just take a generic `SortedRange` and just assume it's sorted the way the user wants.  This function currently won't compile with a non-string `less` lambda.

quickfur replied:
Yes, we will probably have to go this route. I can't think of any sane way to make this function return a meaningful value on an arbitrary predicate.

JackStouffer replied:
Your right, the quantile of a a > b range makes no sense.

quickfur commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4284,247 @@ unittest
+        {
+            ElementType!R prev_item;
+
+            r = r.drop(j_int - 1);
```
Probably should `assert(!r.empty)` here?

JackStouffer replied:
Why? There is already a condition at the start to check if the range is empty.

burner replied:
to check if

prev_item = r.front; 

has actually an element left

JackStouffer replied:
From this equation, it's not possible that there wouldn't be an item left, as `j` will always be smaller than `r.length`. Even when the cast stores `j_int` as 1, the `j_int - 1` won't drop anything.

quickfur commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4284,247 @@ unittest
+{
+    @safe @nogc unittest
+    {
+        assert([1, 2][0 .. 0].assumeSorted.median == double.nan);
```
I don't like this. A function should not have different behaviour based on whether it's compiled in release mode or not. If you think that `median` should return NaN for empty ranges, then it should return NaN for empty ranges, not do one thing in release mode, and something else in non-release mode. If you think it should assert, then it should assert (and leave release mode to UB, because a broken assertion means UB in release mode). Either that, or it should `assert(0)` with an appropriate error message.

Either way, the function should be consistent. It shouldn't assert in non-release mode yet return NaN in release mode.

quickfur commented:
Also, probably should also unittest the corner case of `median` with a single-element range. Does it do the right thing in that case?

yebblies commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4284,247 @@ unittest
+        return r.front;
+    }
+
+    if (less.countUntil(">") > -1)
```
Because of this.  There is no guarantee that `less` will be a `string`.

quickfur replied:
Yeah, this line is extremely scary, because a user-provided `less` predicate, even if it's a string, may not have any `<` or `>` symbol in it. E.g., if it's a custom float type or some such, and the predicate calls an internal function for comparisons, e.g., `a.myFloatType.util.isLessThan(b)`.

JackStouffer commented:
Adressed all concerns.

Not only is this function not pure because of modf, but it's also not CTFE-able, which significantly reduces it's usefulness IMO. Let's hope Iain Buclaw's math refactorings come sooner rather than later.

yebblies commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4284,222 @@ unittest
+Returns:
+    the quantile of the range given $(D q)
+*/
+double quantile(R)(SortedRange!(R, "a < b") r, double q)
```
This isn't what I meant.  It should be (not the actual syntax)
`double quantile(R : SortedRange)(R r, double q)`.
It should accept predicates other than less, for example `(a, b) => a < b` should be fine.  There is no way to detect this, we just have to document it.

JackStouffer replied:
I'm not quite sure how to do this, as `is (R == SortedRange)` returns an error

```
struct std.range.SortedRange(Range, alias pred = "a < b") if (isInputRange!Range) is used as a type
```

yebblies replied:
Something like `is(R : SortedRange!(X, P), X, P)` works I think.  There may be a wrapper for this in std.traits.

JakobOvrum commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4284,222 @@ unittest
+{
+    static if (hasLength!R)
+    {
+        if (r.length == 0)
```
Use `r.empty` instead, since it catches cases where `!hasLength!R`.

JakobOvrum commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4284,222 @@ unittest
+}
+
+///
+@safe pure nothrow unittest
```
You can make the tests `@nogc` by statically allocating the test data:
```d
static immutable arr = [1, 2, 3];
assert(mean(arr[]) == 2);
```

JackStouffer replied:
Unfortunately, `assumeSorted` is not @nogc in debug mode https://issues.dlang.org/show_bug.cgi?id=15003.

JackStouffer commented:
Adressed all concerns and rebased.

JackStouffer commented:
Ping? I have another PR waiting on this.

Also, for some reason druntime unit tests on 32bit linux are failing. I doubt that has anything to do with this PR.

burner commented:
@JackStouffer not everything has been addressed.

please recheck

JackStouffer commented:
Changed the return types of the two functions to always return `real`s, as having the mean of `[1, 2]` as `1` makes no sense.

JackStouffer commented:
Druntime failed again :(

burner commented:
Please test with an empty range, then it LGTM

JackStouffer commented:
Done.

burner commented:
LGTM

quickfur commented:
What happened to the latest comments from myself and @yebblies ? Thought I saw them earlier but can't find them now...

JackStouffer commented:
Other than the comments made by yebblies five days ago on the SortedRange issue, I never received email notifications about any comments.

JackStouffer commented:
@quickfur @DmitryOlshansky This got an OK from @burner, can I get your opinions.

JakobOvrum commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4284,218 @@ unittest
+    {
+        import std.typecons : tuple;
+
+        auto pair = reduce!((a, b) => tuple(a[0] + 1, a[1] + b))
```
It would be nice if we could have `sum`'s accurate summation for the non-length case too - it does support it - but I understand why that's not practical right now. Just wanted to point out that this path is less accurate.

JackStouffer replied:
The only way I see to do that is to modify `sum` to allow for an optional `ref length` parameter, and that seems like a lot of work for little gain.

JakobOvrum replied:
It's an issue of reductive algorithms in general. The information about how many iterations were performed is lost, but it scales poorly to modify each generic algorithm to optionally retain the number of iterations.

Maybe the solution is a higher order range that wraps an input range and tracks the number of pops. After being passed to a reductive algorithm it would result in presenting the effective length of the underlying range. I'm not sure how it should interact with reference-ness and `save`.

In regards to this PR, maybe it's appropriate to leave a comment explicitly stating the accuracy difference between the two paths.

JackStouffer replied:
Updated docs per your suggestion.

John-Colvin replied:
If you got rid of the `static assert (isFloatingPoint!Result);` from `sumPairwise` you could do this:
```D
struct Pair(E)
{
    E e;
    size_t i;
    auto opBinary(string op : "+")(Pair!E rhs)
    {
        return Pair!E(e + rhs.e, i + rhs.i);
    }
}
auto sAndT = zip(repeat(1), r).map!(p => Pair!(ElementType!R)(p.expand)).sumPairwise();
return sAndT.e / sAndT.i;
```

But that's hardly ideal itself and to be honest I think it's OK as you have it already, as long as it's documented correctly.

JackStouffer commented:
All tests are green and no one seems to have any more objections.

JackStouffer commented:
ping @DmitryOlshansky @JakobOvrum @quickfur

JackStouffer commented:
@John-Colvin can I get your opinion on this, seeing as you're the unofficial D statistics guy.

JackStouffer commented:
Ok, now that's three people who have given this PR an ok. Ready to merge.

JackStouffer commented:
Ping @DmitryOlshansky @JakobOvrum @quickfur @schveiguy 

It has been 11 days since this has been ready to merge (6 if you count very minor doc updates). As I said before I have another PR waiting on this, it would be great to get this merged soon.

JakobOvrum commented:
Overall it LGTM, it's just that I'm not really qualified to spot issues in the quantile algorithm.

JackStouffer commented:
@JakobOvrum Yes, but burner and John-Colvin, who seem confident in their knowledge of these algorithms, gave it an ok.

John-Colvin commented:
@JackStouffer That's not really a fair representation of what I've said. To be clear: I think `mean` is ok. I haven't reviewed `quantile` but will do so now.

JackStouffer commented:
@John-Colvin Sorry, I took your silence as acceptance.

burner commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4286,220 @@ unittest
+        !isInfinite!R)
+in
+{
+    assert(q > 0 && q < 1);
```
I really don't like that assert, it is the same as assert(!e.empty)

JackStouffer replied:
> it is the same as assert(!e.empty)

Does that mean I should move it to the body? Or that you don't like its existence?

burner replied:
no return NaN

JackStouffer replied:
> no return NaN

Are you sure this function should do that? Because having a `q` that's less than zero or greater than one makes no sense what so ever, and I feel that accepting it anyway without error is wrong.

burner replied:
and create a function

```D
bool isValidQuantil(real q);
```

JackStouffer replied:
What? Why? This is the type of thing that contracts are made for. Why not use them?

burner replied:
because the compiler can remove them, without you knowing.
and this way it is more composable

JackStouffer replied:
> because the compiler can remove them, without you knowing.

And that's fine, because they'll be in release mode which must be explicitly activated and they should have run it with contracts on at least once while developing. This seems to be perfectly acceptable for a lot of the functions in other modules, especially `std.range` which makes heavy use of contracts.

John-Colvin commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4286,220 @@ unittest
+    real x = r.length * q;
+    real j;
+    real g = modf(x, j);
+    uint j_int = cast(uint) j;
```
should be `size_t`

John-Colvin commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4286,220 @@ unittest
+*/
+real quantile(R)(R r, real q)
+    if (isInstanceOf!(SortedRange, R) &&
+        is(ElementType!R : real) &&
```
Why require this? Surely it's possible to calculate quantiles for any element type.

JackStouffer replied:
> Why require this? Surely it's possible to calculate quantiles for any element type

Do you mean for any built-in type? Because it works for all numerical types. Or do you mean user defined types?

> The same goes for the return value

It's `real` because this makes no sense
```
assert([1, 2, 3, 4].quantile(.5) == 2);
```

John-Colvin commented on std/algorithm/iteration.d:
```diff
@@ -4280,3 +4286,220 @@ unittest
+    {
+        if (g.approxEqual(0))
+        {
+            return (r[j_int - 1] + r[j_int]) / 2.0;
```
Ok, so you take the mean if you're sitting halfway between points. E.g. median of [0, 1, 2, 3] can be defined to be (1 + 2)/2 = 1.5

You haven't covered all the other situations where you fall some other fractional distance between points. It seems like you're generalising the "take the average" approach from calculating medians.

JackStouffer replied:
Look's like I was going off of a rather [simplistic quantile calculation](http://www.stat.wisc.edu/~wardrop/courses/chap12.pdf). I found a [better one](https://www.stat.auckland.ac.nz/~ihaka/787/lectures-quantiles-handouts.pdf) that uses linear interpolation.

burner commented:
@JackStouffer on a meta note.

I think you pushing to hard. D is a "open source/no one is paid to work" project. Asking people to do work for you drives people away, and they do less. This comes from personal experience. 

What makes me really itchy though is:
> Ok, now that's three people who have given this PR an ok. Ready to merge.
when ctrl-f "LGTM" only today gives me two people actually saving LGTM. when D has as signed-off-by then it is LGTM. 

This functionally will get in eventually, but not as fast as you like

JackStouffer commented:
Sorry, you're right. 

I was frustrated that people had given an ok but it was just sitting here. I'm happy to fix things if people tell me they're broken, I just didn't want to see this die on the vine.

burner commented:
it will not die. just use more of the number one computer science skill:
**frustration tolerance**

John-Colvin commented:
Quantiles from data aren't really a trivial, done-and-dusted thing. See https://stat.ethz.ch/R-manual/R-devel/library/stats/html/quantile.html for a serious quantile function and the wikipedia article for an example of a simple one.

A simple quantile, like is shown in the wikipedia example can be done something like this: `auto idx = size_t(round(r.length * q)); r.topN(idx); return i[idx];` (although obviously you wouldn't want to do that sort for every call to quantile, I just wanted to illustrate how easy it can be).

Overall, I'm not sure something like `quantile` belongs in `std.algorithm`, especially if it is going to do any interpolation (i.e. anything other than "choose the nearest index"). It's too much messy statistics decisions, not enough generic manipulation of arbitrary data. If your not doing any interpolation though, it's no longer really much of an algorithm, it's just "read the value at the index closest to q*length, assuming sortedness".

Could I suggest that if you're interested in doing a more comprehensive treatment of quantiles, you could make a pull request to https://github.com/DlangScience/dstats with it?

`mean` is a safer proposition, but it'd be great to hear from @andralex about whether he sees this as within the remit of `std.algorithm` or whether it should go elsewhere. The thing is, it's not very generic, seeing as it only works (by definition), with types that support addition and division.

JackStouffer commented:
> Quantiles from data aren't really a trivial, done-and-dusted thing. See https://stat.ethz.ch/R-manual/R-devel/library/stats/html/quantile.html for a serious quantile function and the wikipedia article for an example of a simple one.

Thanks for the link.

> The thing is, it's not very generic, seeing as it only works (by definition), with types that emulate real numbers.

I'll make the functions more generic by allowing any user defined type that defines `+` and `/` and test with `BigInt`.

> but it'd be great to hear from @andralex about whether he sees this as within the remit of std.algorithm or whether it should go elsewhere (e.g. std.math?).

I don't think std.math is a good fit, as there are no other range based functions in there. I think the only other module that fits is std.numeric, but that seems to be an attempt at a strait one to one implementation of the STL.

> Could I suggest that if you're interested in doing a more comprehensive treatment of quantiles, you could make a pull request to https://github.com/DlangScience/dstats with it?

I keep quantile in there for now, until more people in this PR tell me it's no good for Phobos. If that's true, I'll open a PR on that repo.

JackStouffer commented:
After doing more research, I have concluded that the `quantile` is not nearly complex enough to be useful and requires a much more through look. So, I removed it from this PR, I added the mean overloads back in, and I added support for user defined types to both functions that don't implicitly convert to `real`.

I'll open a WIP PR on dstats and if it's good it can always be added back into phobos later.

quickfur commented:
I think `mean` is ready to go (has been ready for a while now). Perhaps the thing to do here is to split this PR into one for `mean` and one for `median`/`quantile`/etc.. The first one can probably be merged already, it's the second one that's been holding this up.

Generally, it's a good idea to split things up into multiple PRs when there are various pieces that don't directly depend on each other, otherwise perfectly good changes will get held up unnecessarily just because another, mostly-unrelated piece.

JackStouffer commented:
@quickfur ok, but before I do that there was one thing that I have been thinking about with this PR.

I also want to add various other statistical methods to Phobos, such as standard deviation, which fit in `std.algorithm`, and things such as rolling mean, which don't fit into `std.algorithm`. I was thinking that something like a rolling mean would best fit in `std.numeric`, but, would it be best if all of the statistical methods were in the same module, or is it ok to have them separate?

John-Colvin commented:
@JackStouffer this is relevant to discussions about rolling means etc. https://github.com/D-Programming-Language/phobos/pull/2991

JackStouffer commented:
@John-Colvin looks interesting. I'll wait on the rolling mean/median stuff until a decision is reached on that PR. In the mean (ha) time, `mean` and `median` need to have a definite home.

JackStouffer commented:
https://github.com/D-Programming-Language/phobos/pull/3679
https://github.com/D-Programming-Language/phobos/pull/3680
